### DISCUSSION TITLE
The reason why Nuxt runs "lazy" fetch/asyncData on `beforeMount` and not right after `create`

### DISCUSSION THREAD

**Discussion by enkot** [nuxt/nuxt#18520] [Questions] [answered]
It looks like that this:
```ts
const { data } = useFetch('...', { lazy: true })
```
and this:
```ts
const { data } = useFetch('...')
``` 
behave the same way.

Do we really need `lazy` prop or `useLazyFetch` composable? Because, looks like, `useFetch` works exactly the same as `useLazyFetch`. In SSR it properly renders data even without `lazy`.

Based on `useAsyncData` source code, what `lazy` do in CSR - it runs the fetch `onBeforeMount` and not immediately: https://github.com/nuxt/nuxt/blob/1211f62d41b54592dd3bc0d9db62ef9ca0296177/packages/nuxt/src/app/composables/asyncData.ts#L217-L224
But for SSR it works the same for lazy and non-lazy fetch:
https://github.com/nuxt/nuxt/blob/1211f62d41b54592dd3bc0d9db62ef9ca0296177/packages/nuxt/src/app/composables/asyncData.ts#L189-L196

**Comment by AndrewBogdanovTSS** [comment]
My understanding is useFetch should always be used with await, but for useLazyFetch **await** is optional, so I would rather formulate the above question as:

 Does this:
```
const { data } = await useLazyFetch('...')
```
and this:
```
const { data } = useFetch('...')
```
behave the same way?

**Comment by avinean** [comment]
Obviously both composables behave not the same according to docs https://nuxt.com/docs/api/composables/use-lazy-fetch.
`useFetch` - blocks navigation
`useLazyFetch/useFetch('',{lazy: true})` - doesn't block navigation
It means that with `useFetch` you'll never see page until all data fetches.
While with `useLazyFetch/useFetch('',{lazy: true})` you'll see page even if no data has fetched.

But question regarding `await` still remains. 
And here is an answer https://vuejs.org/api/sfc-script-setup.html#top-level-await.
When you use top-level await you `setup` transform to `async setup` and rendering waits until promise resolves.

**Reply by AndrewBogdanovTSS** [reply]
Does it mean that lazy: true in combination with await does nothing?

**Reply by avinean** [reply]
Looks like that.

**Comment by avinean** [comment]
And I'm not sure if title is correct `Does useFetch without await behave like useLazyFetch?`
How `await` keyword can change internal behaviour of a function?

**Reply by enkot** [reply]
@avinean Changed the title. Thanks!

**Comment by guilledll** [comment]
## 🎓   Differences I found on my real world App
After some attempts I finally found how `useFetch`, `useLazyFetch`, `useAsyncData` and `useLazyAsyncData` differ from each other.

For this we have to consider two scenarios, lets asume we are loading a page:

1. Loading a page from **SSR** (direct hit from the URL)
2. Loading a page from **CSR** (normal navigation)

## CSR
While navigating composables will behave this way:

### Using "await"
```js
// Since we are awaiting both "useFetch" and "useAsyncData"
// will block navigation until the promise is resolved (show loading bar)
// resulting "data" will hold the resolved value

const {data} = await useFetch('...');
// Same as doing ...
const {data} = await useAsyncData(() => {});
```

### NOT using "await"
```js
// In case we do not use "await" on any of both we must handle pending state
// this is because "data" will be null until promise is resolved internally
// this does NOT show loading bar <--

// Need to control pending since data can be "null" at first
const {data, pending} = useFetch('...');
// Same for this case...
const {data, pending} = useAsyncData(() => {});
```
### Understanding "lazy" composables
```js
// When using "lazy" composables the "await" keyword is REDUNDANT
// adding it will not change anything during the execution
// since both composables are designed to DO NOT block navigation (does NOT show loading bar)

//                        | ---- > await is REDUNDANT here
const {data, pending} = await useLazyFetch('...');
// Same as doing ...
//                        | ---- > await is REDUNDANT here too
const {data, pending} = await useLazyAsyncData(() => {});

// So ABOVE is equivalent of doing ...
const {data, pending} = useLazyFetch('...');
const {data, pending} = useLazyAsyncData(() => {});
```

### SSR
On the other hand since SSR is resolved on the server all functions will **return the processed value** on the first page render.
"data" will be resolved so there is **no need to handle pending state** during SSR.

In **conclusion** during SSR

✅ `useFetch("...")` === `useLazyFetch("...")` === `useAsyncData(() => {})` === `useLazyAsyncData(() => {})` ✅ 

Hope it helps @enkot @avinean @AndrewBogdanovTSS

**Reply by AndrewBogdanovTSS** [reply]
what about `useFetch(data, {lazy: true})`?

**Comment by enkot** [comment, marked as answer]
Sorry, question was asked incorrectly. 
More correct: 
**What is the reason Nuxt runs "lazy" fetch/asyncData on `beforeMount` and not right after `create`**
here -> https://github.com/nuxt/nuxt/blob/main/packages/nuxt/src/app/composables/asyncData.ts#L257

@danielroe already answered it:
**The reason is that it can cause a hydration mismatch if the fetch 'resolves' before the component has finished hydrating.**

_Originally posted by @danielroe in https://github.com/nuxt/nuxt/issues/22682#issuecomment-1683992594_
